src/libopensc/pkcs15.c

```diff
@@ -1503,19 +1484,15 @@ compare_obj_path(struct sc_pkcs15_object *obj, const struct sc_path *path)
 {
 	void *data = obj->data;
 
-	switch (obj->type) {
-	case SC_PKCS15_TYPE_CERT_X509:
-		return sc_compare_path(&((struct sc_pkcs15_cert_info *) data)->path, path);
-	case SC_PKCS15_TYPE_PRKEY_RSA:
-	case SC_PKCS15_TYPE_PRKEY_DSA:
-	case SC_PKCS15_TYPE_PRKEY_GOSTR3410:
-	case SC_PKCS15_TYPE_PRKEY_EC:
+	switch (obj->type & SC_PKCS15_TYPE_CLASS_MASK) {
+	case SC_PKCS15_TYPE_PRKEY:
 		return sc_compare_path(&((struct sc_pkcs15_prkey_info *) data)->path, path);
-	case SC_PKCS15_TYPE_PUBKEY_RSA:
-	case SC_PKCS15_TYPE_PUBKEY_DSA:
-	case SC_PKCS15_TYPE_PUBKEY_GOSTR3410:
-	case SC_PKCS15_TYPE_PUBKEY_EC:
+	case SC_PKCS15_TYPE_PUBKEY:
 		return sc_compare_path(&((struct sc_pkcs15_pubkey_info *) data)->path, path);
+	case SC_PKCS15_TYPE_SKEY:
+		return sc_compare_path(&((struct sc_pkcs15_prkey_info *) data)->path, path);
```

Comment:
It has to be rather:
`+		return sc_compare_path(&((struct sc_pkcs15_skey_info *) data)->path, path);`